crates/amaru-ledger/src/summary/governance.rs

```diff
@@ -40,57 +43,59 @@ pub struct ProposalState {
 }
 
-impl GovernanceSummary {
-    pub fn new(db: &impl Snapshot) -> Result<Self, StoreError> {
-        let epoch = db.epoch();
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("time horizon error: {0}")]
+    TimeHorizonError(u64, TimeHorizonError),
+    #[error("store error: {0}")]
+    StoreError(#[from] StoreError),
+}
 
-        let mut all_proposals_epochs = BTreeSet::new();
+impl GovernanceSummary {
+    pub fn new(db: &impl Snapshot, era_history: &EraHistory) -> Result<Self, Error> {
+        let mut proposals = BTreeSet::new();
 
         // FIXME: filter out proposals that have been ratified
-        let deposits = db
-            .iter_proposals()?
-            .filter_map(|(_, row)| {
-                all_proposals_epochs.insert(row.proposed_in);
+        let mut deposits = BTreeMap::new();
+
+        db.iter_proposals()?
+            .try_for_each(|(_, row)| -> Result<(), Error> {
+                let epoch = era_history
+                    .slot_to_epoch(Slot::from(row.proposed_in.transaction.slot))
+                    .map_err(|e| Error::TimeHorizonError(row.proposed_in.transaction.slot, e))?;
+
+                proposals.insert((row.proposed_in.transaction, epoch));
+
                 // NOTE: Proposals are ratified with an epoch of delay always, so deposits count
                 // towards the voting stake of DRep for an extra epoch following the proposal
                 // expiry.
                 if epoch <= row.valid_until + 1 {
-                    Some((
+                    deposits.insert(
                         expect_stake_credential(&row.proposal.reward_account),
                         ProposalState {
                             deposit: row.proposal.deposit,
                             valid_until: row.valid_until,
                         },
-                    ))
-                } else {
-                    None
+                    );
                 }
-            })
-            .collect::<BTreeMap<_, _>>();
-
-        let current_epoch = db.epoch();
-
-        // A set containing all overlapping activity periods of all proposals. Might contain disjoint periods.
-        // e.g.
-        //
-        // Considering a proposal created at epoch 163, it is valid until epoch 163 + GOV_ACTION_LIFETIME
-        //
-        // for epochs 163 and 165, with GOV_ACTION_LIFETIME = 6, proposals_activity_periods would equal
-        //   [163, 164, 165, 166, 167, 168, 169, 170, 171]
-        //
-        // for epochs 163 and 172, with GOV_ACTION_LIFETIME = 6, proposals_activity_periods would equal
-        //   [163, 164, 165, 166, 167, 168, 169, 172, 173, 174, 175, 176, 177, 178]
-        let proposals_activity_periods = all_proposals_epochs
-            .iter()
-            .flat_map(|&value| (value..=value + GOV_ACTION_LIFETIME).collect::<BTreeSet<_>>())
-            .collect::<BTreeSet<u64>>();
+
+                Ok(())
+            })?;
+
+        let mandate = drep_mandate_calculator(
```

Comment:
Do we want to invest more energy in having a clean generic mechanism for this? Assuming there will be more such cases and potential future new hard forks.

Reply:
I mean having some specific architecture to handle protocol version specific logic so that it's easy to spot those. As opposed to have those logic scattered in the code base.